pkg/obfuscate/shell_lexer.go

```diff
@@ -0,0 +1,395 @@
+	// Handle double quoted strings
+	if token.Kind == DoubleQuote {
+		if state.current == DoubleQuote {
+			state.current = 0
+		} else {
+			state.current = DoubleQuote
+		}
+	}
+
+	// Handle single quoted strings
+	if token.Kind == SingleQuote {
+		if state.current == SingleQuote {
+			state.current = 0
+		} else {
+			state.current = SingleQuote
+		}
+	}
+
+	// Handle shell subcommands $() in double quoted strings
+	if token.Kind == Field && len(ret) > 0 && ret[len(ret)-1].Kind == DoubleQuote {
+		reg := regexp.MustCompile(`\$\(`)
+		matches := reg.FindAllStringSubmatchIndex(token.Val, -1)
+
+		endOfLastSubcommand := 0
+		for _, match := range matches {
+			// Don't take match if it inside another subcommand that has already been tokenized
+			if match[0] < endOfLastSubcommand {
+				continue
+			}
+
+			dollarCharIndex := match[0]
+			closingParenthesisIndex := getClosingParenthesisIndex(token.Val, dollarCharIndex+1)
+			if closingParenthesisIndex == -1 {
+				// Did not find closing parentheses
+				continue
+			}
+
+			// We now need to split the existing token into a maximum of 3 parts
+			// The part before the subcommand (field), the subcommand itself (list of tokens) and the part after the subcommand (field)
+
+			if dollarCharIndex > endOfLastSubcommand {
+				// There is a part before the subcommand
+				previous := token.Val[endOfLastSubcommand:dollarCharIndex]
+				preToken := ShellToken{Field, previous, token.Start + endOfLastSubcommand, token.Start + dollarCharIndex}
+				ret = append(ret, preToken)
+			}
+
+			// Tokenize shell subcommand (without fully parsing it and change the states of tokens)
+			subCmdTokens := tokenizeShell(token.Val[dollarCharIndex:closingParenthesisIndex])
+
+			// Dummy array for applying real changes
+			// Update offsets (start, end) of subcommand tokens
+			var finalSubCmdTokens []ShellToken
+
+			offset := token.Start + dollarCharIndex
+			for _, tok := range subCmdTokens {
+				tok.Start += offset
+				tok.End += offset
+
+				// append
+				finalSubCmdTokens = append(finalSubCmdTokens, tok)
+			}
+
+			ret = append(ret, finalSubCmdTokens...)
+			endOfLastSubcommand = closingParenthesisIndex
+		}
+
+		if endOfLastSubcommand < len(token.Val)-1 {
+			// There is a part after the last subcommand
+			after := token.Val[endOfLastSubcommand:]
+			postToken := ShellToken{Field, after, token.Start + endOfLastSubcommand, token.End}
+			ret = append(ret, postToken)
+		}
+	} else {
+		ret = append(ret, *token)
+	}
+
+	return ret
+}
+
+func getClosingParenthesisIndex(str string, start int) int {
+	open := 1
+	for i := start + 1; i < len(str); i++ {
+		if str[i] == '(' {
+			open++
+		} else if str[i] == ')' {
+			open--
+		}
+
+		if open == 0 {
+			return i + 1
+		}
+	}
+
+	return -1
+}
+
+// changeStates changes tokens kind based on their context and state
+// remove whitespaces and set executable tokens
+//
+//revive:disable-line:empty-block
```

Comment:
I might be mis-understanding this revive command but I don't see any empty blocks here 🤔 is this still needed?